hail/python/hail/expr/aggregators/aggregators.py

```diff
@@ -184,11 +184,22 @@ def array_agg(self, array, f):
                               Indices(indices.source, aggregated._indices.axes),
                               aggregations)
 
+    @property
+    def context(self):
+        if self._as_scan:
+            return 'scan'
+        else:
+            return 'agg'
+
 _agg_func = AggFunc()
 
 
 def _check_agg_bindings(expr, bindings):
-    bound_references = {ref.name for ref in expr._ir.search(lambda ir: isinstance(ir, Ref) and not isinstance(ir, TopLevelReference))}
+    bound_references = {ref.name for ref in expr._ir.search(
+        lambda ir: isinstance(ir, Ref)
+                   and not isinstance(ir, TopLevelReference)
+                   and not ir.name.startswith('__uid_scan')
+                   and not ir.name.startswith('__uid_agg'))}
     free_variables = bound_references - expr._ir.bound_variables - bindings
```

Comment:
Ah right, now I remember why this is wrong. In something like
```
Ref('i') + ArrayMap(a, 'i', body)
```
`i` is free, but this would find no free variables. I can fix this as part of cleaning up the type and context handling in python.